src/webui/WebUI.jl

```diff
@@ -0,0 +1,227 @@
+    @error "Handler error" route=r.target
+    html(500, "Server error, sorry!")
+end
+
+pathmatch(p::AbstractString, f::Function) = branch(r -> first(split(r.target, "?")) == p, f)
+
+function action(regdata::RegistrationData, zsock::RequestSocket)
+    regp = RegEdit.RegisterParams(cloneurl(regdata.repo), regdata.project, regdata.tree;
+        registry=REGISTRY[].clone, registry_deps=REGISTRY[].deps, push=true,
+        gitconfig=Dict("user.name"=>CONFIG["user"], "user.email"=>CONFIG["email"]),
+    )
+    branch = sendrecv(zsock, regp; nretry=5)
+    if branch === nothing || get(branch.metadata, "error", nothing) !== nothing
+        if branch === nothing
+            msg = "ERROR: Registrator backend service unreachable"
+        else
+            msg = "Registration failed: " * branch.metadata["error"]
+        end
+        state = :errored
+    else
+        title, body = pull_request_contents(;
+            registration_type=get(branch.metadata, "kind", ""),
+            package=regdata.project.name,
+            repo=web_url(regdata.repo),
+            user=display_user(regdata.user),
+            gitref=regdata.ref,
+            version=regdata.project.version,
+            commit=regdata.commit,
+            patch_notes=regdata.notes,
+        )
+
+        # Make the PR.
+        pr = @gf make_registration_request(REGISTRY[], branch.branch, title, body)
+        if pr === nothing
+            msg = "Registration failed: Making pull request failed"
+            state = :errored
+        else
+            url = web_url(pr)
+            msg = """Registry PR successfully created, see it <a href="$url" target="_blank">here</a>!"""
+            state = :success
+        end
+    end
+    @debug msg
+    REGISTRATIONS[regdata.commit] = RegistrationState(msg, state)
+    nothing
+end
+
+function request_processor(zsock::RequestSocket)
+    do_action() = action(take!(event_queue), zsock)
+    handle_exception(ex) = (isa(ex, InvalidStateException) && (ex.state == :closed)) ? :exit : :continue
+    keep_running() = isopen(httpsock[])
+    recover("request_processor", keep_running, do_action, handle_exception)
+end
+
+function start_server(ip::IPAddr, port::Int)
+    httpsock[] = Sockets.listen(ip, port)
+    @app server = (
+        error_handler,
+        pathmatch(ROUTE_INDEX, index),
+        pathmatch(ROUTE_AUTH, auth),
+        pathmatch(ROUTE_CALLBACK, callback),
+        pathmatch(ROUTE_SELECT, select),
+        pathmatch(ROUTE_REGISTER, register),
+        pathmatch(ROUTE_STATUS, status),
+        r -> html(404, "Page not found"),
+    )
+    do_action() = begin; t = serve(server, ip, port; server=httpsock[], readtimeout=0); wait(t); end
+    handle_exception(ex) = (isa(ex, Base.IOError) && (ex.code == -103)) ? :exit : :continue
+    keep_running() = isopen(httpsock[])
+    @info "Serving" ip port
+    recover("webui", keep_running, do_action, handle_exception)
+end
+
+function main(config::AbstractString="config.toml")
```

Comment:
This should probably be something along the lines of:

```jl
function main(config::AbstractString=isempty(ARGS) ? "config.toml" : first(ARGS))
```
It's most intuitive if the server dies immediately for non-existent files.